compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -2585,7 +2585,7 @@ impl<'tcx> TyCtxt<'tcx> {
         self.fn_sig(trait_item_def_id).subst_identity().skip_binder().output().walk().any(|arg| {
             if let ty::GenericArgKind::Type(ty) = arg.unpack()
                 && let ty::Alias(ty::Projection, data) = ty.kind()
-                && self.def_kind(data.def_id) == DefKind::ImplTraitPlaceholder
+                && self.is_impl_trait_in_trait(data.def_id)
```

Comment:
This fn could just call `associated_items_for_impl_trait_in_trait(trait_item_def_id).is_empty()` when the `-Z` flag is true, instead of walking the signature.